lib/auth/oidc_test.go

```diff
@@ -782,3 +782,104 @@ func TestEmailVerifiedClaim(t *testing.T) {
+		Allow: types.RoleConditions{},
+	})
+	require.NoError(t, err)
+	err = s.a.CreateRole(ctx, role)
+	require.NoError(t, err)
+
+	// Create claims with "preferred_username" field.
+	claims := map[string]interface{}{
+		"email_verified":     true,
+		"groups":             []string{"everyone"},
+		"email":              "test-user@example.com",
+		"sub":                "00001234abcd",
+		"exp":                1652091713.0,
+		"preferred_username": "Teleport_TestUser",
+	}
+
+	// Create identity from the claims.
+	ident, err := oidc.IdentityFromClaims(claims)
+	require.NoError(t, err)
+
+	tests := []struct {
+		desc             string
+		spec             types.OIDCConnectorSpecV3
+		expectedUsername string
+		expectedError    string
+	}{
+		{
+			desc: "username_claim specified with correct claim",
+			spec: types.OIDCConnectorSpecV3{
+				IssuerURL:     idp.s.URL,
+				ClientID:      "000",
+				ClientSecret:  "0000",
+				ClaimsToRoles: []types.ClaimMapping{{Claim: "groups", Value: "everyone", Roles: []string{"access"}}},
+				RedirectURLs:  []string{"https://proxy.example.com/v1/webapi/oidc/callback"},
+				UsernameClaim: "preferred_username",
+			},
+			expectedUsername: "Teleport_TestUser",
+		},
+		{
+			desc: "username_claim specified with incorrect claim",
+			spec: types.OIDCConnectorSpecV3{
+				IssuerURL:     idp.s.URL,
+				ClientID:      "000",
+				ClientSecret:  "0000",
+				ClaimsToRoles: []types.ClaimMapping{{Claim: "groups", Value: "everyone", Roles: []string{"access"}}},
+				RedirectURLs:  []string{"https://proxy.example.com/v1/webapi/oidc/callback"},
+				UsernameClaim: "prefred_usrnam",
+			},
+			expectedError: "The configured username_claim of 'prefred_usrnam' was not received from the IdP. Please update the username_claim in connector 'okta-oidc'.",
+		},
+		{
+			desc: "no username_claim specified, default to using email",
+			spec: types.OIDCConnectorSpecV3{
+				IssuerURL:     idp.s.URL,
+				ClientID:      "000",
+				ClientSecret:  "0000",
+				ClaimsToRoles: []types.ClaimMapping{{Claim: "groups", Value: "everyone", Roles: []string{"access"}}},
+				RedirectURLs:  []string{"https://proxy.example.com/v1/webapi/oidc/callback"},
+			},
+			expectedUsername: "test-user@example.com",
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.desc, func(t *testing.T) {
+			// Create OIDC connector with UsernameClaim specified.
+			connector, err := types.NewOIDCConnector("okta-oidc", tc.spec)
+
+			// Create OIDC request.
+			oidcRequest := types.OIDCAuthRequest{
+				ConnectorID:   "okta-oidc",
+				Type:          constants.OIDC,
+				CertTTL:       defaults.OIDCAuthRequestTTL,
+				SSOTestFlow:   true,
+				ConnectorSpec: &tc.spec,
+			}
+			request, err := s.a.CreateOIDCAuthRequest(ctx, oidcRequest)
+			require.NoError(t, err)
+
+			// Generate the userCreateParams for the OIDC user.
+			createUserParams, err := s.a.calculateOIDCUser(&diagCtx, connector, claims, ident, request)
+
```

Comment:
Remove empty line here, since the code that follows is directly related to this line.